src/libxtp/davidsonsolver.cc

```diff
@@ -0,0 +1,288 @@
+ * limitations under the License.
+ *
+ */
+
+#include <stdexcept>
+#include <iostream>
+
+#include <Eigen/Dense>
+#include <Eigen/Core>
+#include <Eigen/QR>
+#include <Eigen/Eigenvalues>
+#include <Eigen/IterativeLinearSolvers>
+#include <unsupported/Eigen/IterativeSolvers>
+#include <chrono>
+
+#include <votca/xtp/matrixfreeoperator.h>
+#include <votca/xtp/bse_matrix_free.h>
+#include <votca/xtp/davidsonsolver.h>
+
+using boost::format;
+using std::flush;
+
+namespace votca { namespace xtp {
+
+using namespace std;
+
+DavidsonSolver::DavidsonSolver(ctp::Logger &log) : _log(log), iter_max(1000), tol(1E-6), max_search_space(100) { }
+// DavidsonSolver::DavidsonSolver(int iter_max) : iter_max(iter_max) , tol(1E-6), max_search_space(100) { }
+// DavidsonSolver::DavidsonSolver(int iter_max, real_gwbse tol) : iter_max(iter_max), tol(tol), max_search_space(100) { }
+// DavidsonSolver::DavidsonSolver(int iter_max, real_gwbse tol, int max_search_space) : iter_max(iter_max), tol(tol), max_search_space(max_search_space) { }
+
+void DavidsonSolver::set_iter_max(int N) { this->iter_max = N; }
+void DavidsonSolver::set_tolerance(real_gwbse eps) { this->tol = eps; }
+void DavidsonSolver::set_max_search_space(int N) { this->max_search_space = N;}
+void DavidsonSolver::set_jacobi_correction() { this->jacobi_correction = true; }
+void DavidsonSolver::set_jacobi_linsolve(int method) {this->jacobi_linsolve = method;}
+
+VectorXfd DavidsonSolver::eigenvalues() {return this->_eigenvalues;}
+MatrixXfd DavidsonSolver::eigenvectors() {return this->_eigenvectors;}
+
+
+ArrayXfd DavidsonSolver::_sort_index(VectorXfd V)
+{
+    ArrayXfd idx = ArrayXfd::LinSpaced(V.rows(),0,V.rows()-1);
+    std::sort(idx.data(),idx.data()+idx.size(),
+              [&](int i1, int i2){return V[i1]<V[i2];});
+    return idx; 
+}
+
+MatrixXfd DavidsonSolver::_get_initial_eigenvectors(VectorXfd d, int size_initial_guess)
+{
+
+    MatrixXfd guess = MatrixXfd::Zero(d.size(),size_initial_guess);
+    ArrayXfd idx = DavidsonSolver::_sort_index(d);
+
+    for (int j=0; j<size_initial_guess;j++)
+        guess(idx(j),j) = 1.0;
+
+    return guess;
+}
+
+MatrixXfd DavidsonSolver::_solve_linear_system(MatrixXfd A, VectorXfd r)
+{
+    MatrixXfd w;
+
+    switch(this->jacobi_linsolve)
+    {        
+        //use cg approximate solver     
+        case 0: 
+        {
+            Eigen::ConjugateGradient<MatrixXfd, Eigen::Lower|Eigen::Upper> cg;
+            cg.compute(A);
+            w = cg.solve(r);
+        }   
+
+        //use GMRES approximate solver
+        case 1:
+        {
+            Eigen::GMRES<MatrixXfd, Eigen::IdentityPreconditioner> gmres;
+            gmres.compute(A);
+            w = gmres.solve(r);
+        }
+
+        //use llt exact solver
+        case 2: w = A.llt().solve(r);
+    }
+
+    return w;
+}
+
+
+
+template <class OpMat>
+MatrixXfd DavidsonSolver::_jacobi_orthogonal_correction(OpMat A, VectorXfd r, VectorXfd u, real_gwbse lambda)
```

Comment:
u,r, A, lambda are not good names unless they are explained in a header comment.